### PR TITLE
execution, db: unwind/reorg tests + concurrency and shutdown fixes

### PR DESCRIPTION
**Pull request by yperbasis**
Part of #21860.

Adds integration and stage-level regression tests for the unwind/reorg state-correctness issues tracked in #21860, built around a self-checking, storage-churning contract so that any state left inconsistent by a faulty unwind or reorg surfaces as a failed transaction or a wrong read. The multi-step domain-unwind fix this work uncovered (#21981) is already merged into main.

Production changes: `findExecutedDiffsetAtHeight` is extracted from `unwindExec3` so the #21515 fallback can be unit-tested, plus fixes for two latent infrastructure bugs the new tests surfaced (see Fixes below).

## `StateChurn` contract

A fixed ring of 16 pseudo-random storage keys; each `poke(seed)` overwrites the next key round-robin with `keccak(seed, cursor) % 3`. So ~⅓ of writes set a slot to 0 (deletion) and slots are continually deleted and recreated — the storage pattern that exposes buggy unwinds. `trackedSum` is maintained incrementally and `require`d to equal the sum recomputed from storage.

## Regression tests (red-on-revert — fail without the fix)

- **`TestUnwindExecutionStage_PrunesUncommittedOverlayWrite`** (#21681 / #21824) — when a block fails mid-batch before its step is committed, the no-disk-rollback execution unwind must still prune the uncommitted in-RAM `SharedDomains` overlay to the committed boundary; otherwise the stale writes survive into the unwind→retry loop and re-execution reads a wrong value.
- **`TestFindExecutedDiffsetAtHeight_FallsBackAfterCanonicalReorg`** (#21515) — on a reorg the canonical hash for the unwound range is cleared, so the diffset lookup (keyed on the canonical hash) must fall back to the stored header; otherwise the unwind silently no-ops and the unwound block's state survives as phantom data (→ `CREATE2` collision → gas-used mismatch). Tested at the stage level because the path is only reachable through the full staged-sync pipeline.
- **`TestEngineApiUnwindAcrossDomainStepBoundaries`** (#21981) — full node: builds real domain snapshot files, then unwinds a multi-step range crossing several domain-step boundaries and checks the StateChurn invariant against ground truth. Fails on main without the multi-step-unwind fix.

## Integration coverage (unwind/reorg correctness)

- **`TestEngineApiUnwindRedoStateChurnPreservesState`** — one node records `trackedSum` at every height, then unwinds and redoes across shallow and deep targets, asserting live state matches the recorded value at each. Catches "head moved but state not rolled back".
- **`TestEngineApiReorgToSideChainSwitchesStateChurn`** — reorgs a victim onto a competing side chain that churns different storage, asserting live state switches to the side fork.

## Fixes (latent bugs the new tests surfaced)

- **`db/state` — data race between the background merge and chain-tip worker presets.** The merge path read per-domain compressor/accessor worker counts that chain-tip execution (`PresetChainTipConcurrency`) mutates concurrently; only the collate/build path pinned them. `mergeLoop` now pins worker config for its duration, and the pin is reentrant so an overlapping background build and merge can't re-enable edits mid-operation. (Surfaced by the `-race` run of `TestEngineApiUnwindAcrossDomainStepBoundaries`.)
- **`db/kv/temporal` — the state aggregator was never released at shutdown.** `DB.Close()` closed the underlying MDBX and the forkable aggregators but not the state aggregator it owns, leaving its snapshot-file mmaps open. Harmless on POSIX, but on Windows an open mmap can't be unlinked, so deleting the datadir after shutdown failed. `DB.Close()` now closes the state aggregator too, before the MDBX handle, so its background goroutines release their read transactions first.

### CODE REVIEW

**Comment by yperbasis** [comment]
Validating these StateChurn unwind/reorg tests uncovered a real domain-unwind bug — `DomainRoTx.unwind` leaving stale state when the unwound range crosses a domain-step boundary. Fixed separately in #21981, which also adds a deterministic domain-level regression test (`TestDomain_MultiStepUnwindMatchesGroundTruth`) for the snapshot-file-boundary unwind path. These engine-API tests don't exercise that path (they run on in-DB/recent state; domain-file building in the harness is nondeterministic background work), so the regression guard for it lives at the domain level in #21981.

File: execution/engineapi/engine_api_state_churn_files_test.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
+package engineapi_test
```
[the remark this answers is not in the document]
Comment (yperbasis) [reply execution/engineapi/engine_api_state_churn_files_test.go:17]: Added the standard LGPL header in ef0a7031cf.

File: execution/engineapi/engine_api_state_churn_files_test.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+		Genesis:     genesis,
+		CoinbaseKey: coinbaseKey,
+		EthConfigTweaker: func(c *ethconfig.Config) {
+			c.Snapshot.ProduceE3 = true       // build domain snapshot files
+			c.AlwaysGenerateChangesets = true // keep changesets so the unwind is permitted
+			c.MaxReorgDepth = 90              // below the ~96-block file-boundary lag: unwinds stay above filed state
+		},
+	})
+	require.NoError(t, err)
+	t.Cleanup(func() { _ = eat.Close() })
```
[the remark this answers is not in the document]
Comment (yperbasis) [reply execution/engineapi/engine_api_state_churn_files_test.go]: Fixed in ef0a7031cf — now `require.NoError(t, eat.Close())` to match the other engineapi tests; re-ran the test to confirm Close() succeeds.

File: execution/state/contracts/gen_statechurn.go [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+}
+
+// Poke is a paid mutator transaction binding the contract method 0x32145f90.
+//
+// Solidity: function poke(uint256 seed) returns()
+func (_StateChurn *StateChurnTransactorSession) Poke(seed *big.Int) (types.Transaction, error) {
+	return _StateChurn.Contract.Poke(&_StateChurn.TransactOpts, seed)
+}
+
+// StateChurnPokeParams is an auto generated read-only Go binding of transcaction calldata params
```
[the remark this answers is not in the document]
Comment (yperbasis) [reply execution/state/contracts/gen_statechurn.go:338]: `gen_statechurn.go` is abigen output (`// Code generated by abigen. DO NOT EDIT.`). This `transcaction` typo originates in the generator template (`execution/abi/bind/template.go`) and appears verbatim in every generated binding in the repo, so editing it here would be reverted on the next `go generate` and diverge this binding from the rest. Out of scope for this PR — the proper fix is template-level + regenerate-all.

File: execution/state/contracts/gen_statechurn.go [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+// Solidity: function poke(uint256 seed) returns()
+func ParseStateChurnPokeParams(calldata []byte) (*StateChurnPokeParams, error) {
+	if len(calldata) <= 4 {
+		return nil, fmt.Errorf("invalid calldata input")
+	}
+
+	_abi, err := abi.JSON(strings.NewReader(StateChurnABI))
+	if err != nil {
+		return nil, fmt.Errorf("failed to get abi of registry metadata: %w", err)
+	}
```
[the remark this answers is not in the document]
Comment (yperbasis) [reply execution/state/contracts/gen_statechurn.go:354]: Same here: `gen_statechurn.go` is abigen output (`DO NOT EDIT`), and the `registry metadata` wording is template boilerplate (`execution/abi/bind/template.go`) shared by every generated binding. A manual edit would be lost on regeneration; better addressed at the template level in a separate PR.

**Review by AskAlexSharov — APPROVED** [review]

File: execution/engineapi/engine_api_state_churn_files_test.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+		built := false
+		for _, n := range agg.FilesAmount() {
+			if n > 0 {
+				built = true
+				break
+			}
+		}
+		return built && !agg.HasBackgroundFilesBuild2()
+	}, 60*time.Second, 100*time.Millisecond, "domain files never settled")
+}
```
[the remark this answers is not in the document]
Comment (yperbasis) [reply execution/engineapi/engine_api_state_churn_files_test.go]: Addressed in 92d3e2b5bc: added `require.NotNil(t, agg, "StateAgg is nil: ChainDB did not expose a *state.Aggregator")` at the top of `waitForDomainFilesSettled`, so a failed `StateAgg` assertion fails with a clear message instead of a nil-pointer panic. (The aggregator lookup also now uses the named `state.HasAgg` interface in `InitialiseEngineApiTester`.)

File: execution/engineapi/engine_api_state_churn_reorg_test.go [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+		assertChurnState(ctx, t, eat, churn, sideTip, sideRef)
+	})
+}
+
+// buildChurnChain deploys StateChurn at block 2 on the given tester and then
+// applies one poke per block, using seed(k) for the k-th poke. The returned
+// payloads are indexed by offset from block 2: payloads[0] is the deploy block
+// (height 2) and payloads[i] is the i-th poke block (height 2+i). sums[i] is the
+// on-chain trackedSum observed at payloads[i] and is the ground truth a later
+// unwind/reorg must reproduce.
```
[the remark this answers is not in the document]
Comment (yperbasis) [reply execution/engineapi/engine_api_state_churn_reorg_test.go]: Addressed in 92d3e2b5bc: rewrote the doc comment to be consistent — `payloads[0]` is the deploy block (height 2) and `payloads[k+1]` is the block applying poke k (height 3+k).

File: execution/stagedsync/stage_execute.go [hunk 7]
```diff
@@ -147,6 +147,36 @@ func (cfg ExecuteBlockCfg) WithAuthor(author accounts.Address) ExecuteBlockCfg {
...
+		default:
+			return diffSet, common.Hash{}, false, fmt.Errorf("diffsets ambiguous for: %d, have %d headers", currentBlock, len(nonCanonicalHeaders))
+		}
+	}
+	diffSet, found, err = doms.GetDiffset(rwTx, executedHash, currentBlock)
+	if err != nil {
+		return diffSet, common.Hash{}, false, err
+	}
+	return diffSet, executedHash, found, nil
+}
```
[the remark this answers is not in the document]
Comment (yperbasis) [reply execution/stagedsync/stage_execute.go:178]: Good catch, and you're right that header count is the wrong signal: in a reorg multiple headers can share a height while only the executed block has a stored diffset, so the fallback should select the unique header whose diffset is present (or `found=false` if none) and error only when more than one header actually carries a diffset.

This branch predates the PR — it was moved verbatim from the inline `unwindExec3` — so to keep this PR scoped to the regression tests plus the agreed concurrency/shutdown fixes, I'm tracking the logic fix together with its (currently missing) branch coverage in #22117 rather than changing the diffset-selection behavior here.
